### PR TITLE
SAA-2197: Update attendances for multiple appointments

### PR DESCRIPTION
**Pull request by munslowl**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/appointment/Appointment.kt [hunk 1]
```diff
@@ -203,8 +203,6 @@ data class Appointment(
 
   fun usernames() = listOfNotNull(createdBy, updatedBy, cancelledBy).distinct()
 
-  fun attendeeUsernames() = attendees().flatMap { it.usernames() }.distinct()
-
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/appointment/Appointment.kt:207]: Not used

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/repository/appointment/AppointmentRepository.kt [hunk 2]
```diff
@@ -10,7 +10,7 @@ import java.time.LocalTime
 
 @Repository
 interface AppointmentRepository : JpaRepository<Appointment, Long> {
-  @Query("from Appointment a where a.appointmentId in :ids")
+  @Query("from Appointment a left join fetch a.attendees where a.appointmentId in :ids")
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/repository/appointment/AppointmentRepository.kt:13]: Only ever used by ui for updating attendances and always using the attendees details

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/appointment/AppointmentAttendanceService.kt [hunk 3]
```diff
@@ -130,10 +145,90 @@ class AppointmentAttendanceService(
...
+              AttendanceAction.RESET -> {
+                attendance.attendanceRecordedTime = null
+                attendance.attendanceRecordedBy = null
+                null
+              }
+            }
+
+            attendance.attended = newAttendedState
+
+            if (oldAttendedState != null && oldAttendedState != newAttendedState) {
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/appointment/AppointmentAttendanceService.kt:200]: Only count changes to existing attendances

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/appointment/AppointmentController.kt [hunk 4]
```diff
@@ -424,6 +428,69 @@ class AppointmentController(
...
+    Update the attendance records for attendees on multiple appointments. This sets the current attendance records
+    for the supplied prison numbers, replacing any existing records. This supports both the initial recording of attendances
+    and changing that attendance records. There are no restrictions on when attendances can be recorded. It can be done
+    for past and future appointments.
+    """,
+  )
+  @ApiResponses(
+    value = [
+      ApiResponse(
+        responseCode = "202",
```
Comment (dafhoustonmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/appointment/AppointmentController.kt:446]: I would have thought 200 as it's not async

Comment (munslowl) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/appointment/AppointmentController.kt]: Updated docs and test

**Review by natclamp-moj — APPROVED** [review]

**Review by dafhoustonmoj — APPROVED** [review]

**Pull request merged**